### ISSUE TITLE
Add some rigidity / reference about tables keys

### ISSUE THREAD

**Issue by lcontami** [QUANTECA-QFox/prober_database#4]
Since we allow for dict in the database (for example, measure has a context that could contain "meas_setup": "prober"; wafer contains a list of die sorted by EB_1, EB_2 or DUV depending on insulation), we should ensure that these dict keys are kept coherent. 

How should we do this ?

**Comment by MatthieuDartiailh** [comment]
You could use a named tuple rather than dict to get a rigid structure.

**Comment by lcontami** [comment]
For the database I cannot have any object as an entry. The dict works because it is a JSON-type object.
The second issue is that I would also like to have a convention for values as well as keys.
BUT the whole idea of the json colum was to keep flexibility on the keys :)

We can always wait and see how fast it becomes a mess.

**Comment by lcontami** [comment]
I am sure you are loving this idea ;)

**Comment by MatthieuDartiailh** [comment]
The ORM could perform some automatic conversion no ? If yes you could use atom to type the values at runtime and even specify enumerations for some fields.

**Comment by lcontami** [comment]
Ok thanks @MatthieuDartiailh I have to look into it !

**Comment by MatthieuDartiailh** [comment]
Do you store pickles in the database ?

**Comment by lcontami** [comment]
No, I store my entries locally in a pickle file when the db is not accessible, and when I can connect to the db I store them in a db table with typed columns (there is a JSON-type column that I use for stuff like: setup = {'instru': 'neel', 'machine': prober} ).
Since I use PostGRE I must define a column type and stick to it.

**Comment by MatthieuDartiailh** [comment]
Makes sense

**Comment by lcontami** [comment]
Results from some internet research:

I think JSONB type column is not adapted to our needs:  The JSON type, when used with the SQLAlchemy ORM, does not detect in-place mutations to the structure (probably not bothering for now but still; access should rather be the same for all columns and restricted by user rights)

There are other column type allowing keyword / value seach that I think will also allow typing more easily: HSTORE

In PostGre, there is another column type (HSTORE) that allow to store a dict of str and support filtering by key / value.
Combined with MutableDict, it also allow for in-place mutation
https://www.kite.com/python/docs/sqlalchemy.dialects.postgresql.HSTORE
(ok, mutable can also be used on json encoded dict: MutableDict.as_mutable(JSONEncodedDict))  but can we filter by key / value ?)
To investigate: can I define a new column type that inherits from HSTORE and fix the dict structure ?


Note: then I need to modify the way die column, it is the only one that is not a dict of str.

**Comment by lcontami** [comment]
This pull request is becoming bigger and bigger ! 

It adds a json_columns file that defines key names / value types for certain json columns of the database.
We decided to keep them flexible in the db to reduce the need for migration, at the cost of having to validate against these classes before upload.
